// @klaytn/web3rpc/dist/model/KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _ApiClient = _interopRequireDefault(require("../ApiClient"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /**
 * web3rpc
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.9.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */
/**
 * The KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner model module.
 * @module model/KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner
 * @version 0.9.9
 */
var KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner = /*#__PURE__*/function () {
  /**
   * Constructs a new <code>KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner</code>.
   * @alias module:model/KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner
   */
  function KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner() {
    _classCallCheck(this, KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner);
    KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  _createClass(KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner, null, [{
    key: "initialize",
    value: function initialize(obj) {}

    /**
     * Constructs a <code>KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner</code> from a plain JavaScript object, optionally creating a new instance.
     * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
     * @param {Object} data The plain JavaScript object bearing properties of interest.
     * @param {module:model/KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner} obj Optional instance to populate.
     * @return {module:model/KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner} The populated <code>KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner</code> instance.
     */
  }, {
    key: "constructFromObject",
    value: function constructFromObject(data, obj) {
      if (data) {
        obj = obj || new KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner();
        if (data.hasOwnProperty('V')) {
          obj['V'] = _ApiClient["default"].convertToType(data['V'], 'String');
        }
        if (data.hasOwnProperty('R')) {
          obj['R'] = _ApiClient["default"].convertToType(data['R'], 'String');
        }
        if (data.hasOwnProperty('S')) {
          obj['S'] = _ApiClient["default"].convertToType(data['S'], 'String');
        }
      }
      return obj;
    }

    /**
     * Validates the JSON data with respect to <code>KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner</code>.
     * @param {Object} data The plain JavaScript object bearing properties of interest.
     * @return {boolean} to indicate whether the JSON data is valid with respect to <code>KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner</code>.
     */
  }, {
    key: "validateJSON",
    value: function validateJSON(data) {
      // ensure the json data is a string
      if (data['V'] && !(typeof data['V'] === 'string' || data['V'] instanceof String)) {
        throw new Error("Expected the field `V` to be a primitive type in the JSON string but got " + data['V']);
      }
      // ensure the json data is a string
      if (data['R'] && !(typeof data['R'] === 'string' || data['R'] instanceof String)) {
        throw new Error("Expected the field `R` to be a primitive type in the JSON string but got " + data['R']);
      }
      // ensure the json data is a string
      if (data['S'] && !(typeof data['S'] === 'string' || data['S'] instanceof String)) {
        throw new Error("Expected the field `S` to be a primitive type in the JSON string but got " + data['S']);
      }
      return true;
    }
  }]);
  return KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner;
}();
/**
 * @member {String} V
 */
KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner.prototype['V'] = undefined;

/**
 * @member {String} R
 */
KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner.prototype['R'] = undefined;

/**
 * @member {String} S
 */
KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner.prototype['S'] = undefined;
var _default = exports["default"] = KlayGetBlockWithConsensusInfoByNumberRangeRespResultTransactionsInnerSignaturesInner;
