// @klaytn/web3rpc/dist/model/TracerObject.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _ApiClient = _interopRequireDefault(require("../ApiClient"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); } /**
 * web3rpc
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.9.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 *
 */
/**
 * The TracerObject model module.
 * @module model/TracerObject
 * @version 0.9.9
 */
var TracerObject = /*#__PURE__*/function () {
  /**
   * Constructs a new <code>TracerObject</code>.
   * Tracer type.
   * @alias module:model/TracerObject
   */
  function TracerObject() {
    _classCallCheck(this, TracerObject);
    TracerObject.initialize(this);
  }

  /**
   * Initializes the fields of this object.
   * This method is used by the constructors of any subclasses, in order to implement multiple inheritance (mix-ins).
   * Only for internal use.
   */
  _createClass(TracerObject, null, [{
    key: "initialize",
    value: function initialize(obj) {}

    /**
     * Constructs a <code>TracerObject</code> from a plain JavaScript object, optionally creating a new instance.
     * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
     * @param {Object} data The plain JavaScript object bearing properties of interest.
     * @param {module:model/TracerObject} obj Optional instance to populate.
     * @return {module:model/TracerObject} The populated <code>TracerObject</code> instance.
     */
  }, {
    key: "constructFromObject",
    value: function constructFromObject(data, obj) {
      if (data) {
        obj = obj || new TracerObject();
        if (data.hasOwnProperty('tracer')) {
          obj['tracer'] = _ApiClient["default"].convertToType(data['tracer'], 'String');
        }
      }
      return obj;
    }

    /**
     * Validates the JSON data with respect to <code>TracerObject</code>.
     * @param {Object} data The plain JavaScript object bearing properties of interest.
     * @return {boolean} to indicate whether the JSON data is valid with respect to <code>TracerObject</code>.
     */
  }, {
    key: "validateJSON",
    value: function validateJSON(data) {
      // ensure the json data is a string
      if (data['tracer'] && !(typeof data['tracer'] === 'string' || data['tracer'] instanceof String)) {
        throw new Error("Expected the field `tracer` to be a primitive type in the JSON string but got " + data['tracer']);
      }
      return true;
    }
  }]);
  return TracerObject;
}();
/**
 * Setting this will enable JavaScript-based transaction tracing, described in the next section. If set, the previous four arguments will be ignored. The predefined tracers can also be used as the following table. | Tracer Name | Description | |-----|-----| | 4byteTracer | 4byteTracer searches for 4byte-identifiers, and collects them for post-processing. It collects the methods identifiers along with the size of the supplied data, so a reversed signature can be matched against the size of the data. | | callTracer | callTracer is a full-blown transaction tracer that extracts and reports all the internal calls made by a transaction, along with any useful information. | | fastCallTracer | fastCallTracer is a Go-native version of callTracer. Since it is not executed on JavaScript VM, it shows more than 10x speedup compared to callTracer. Please use fastCallTracer instead of callTracer if the performance is the matter of the first importance. | | evmdisTracer | evmdisTracer returns sufficient information from a trace to perform evmdis-style disassembly. | | noopTracer | noopTracer is just the barebone boilerplate code required from a JavaScript object to be usable as a transaction tracer. | | opcountTracer | opcountTracer is a sample tracer that just counts the number of instructions executed by the KLVM before the transaction terminated. | | prestateTracer | prestateTracer outputs sufficient information to create a local execution of the transaction from a custom assembled genesis block. | | revertTracer | revertTracer outputs the error string of REVERT. If the execution is not reverted, it outputs an empty string. | | unigramTracer | unigramTracer returns the number of occurrences of each opcode. | | bigramTracer | bigramTracer returns the number of occurrences of two consecutive opcodes. | | trigramTracer | trigramTracer returns the number of occurrences of three consecutive opcodes. | 
 * @member {module:model/TracerObject.TracerEnum} tracer
 */
TracerObject.prototype['tracer'] = undefined;

/**
 * Allowed values for the <code>tracer</code> property.
 * @enum {String}
 * @readonly
 */
TracerObject['TracerEnum'] = {
  /**
   * value: "4byteTracer"
   * @const
   */
  "4byteTracer": "4byteTracer",
  /**
   * value: "callTracer"
   * @const
   */
  "callTracer": "callTracer",
  /**
   * value: "fastCallTracer"
   * @const
   */
  "fastCallTracer": "fastCallTracer",
  /**
   * value: "evmdisTracer"
   * @const
   */
  "evmdisTracer": "evmdisTracer",
  /**
   * value: "noopTracer"
   * @const
   */
  "noopTracer": "noopTracer",
  /**
   * value: "opcountTracer"
   * @const
   */
  "opcountTracer": "opcountTracer",
  /**
   * value: "prestateTracer"
   * @const
   */
  "prestateTracer": "prestateTracer",
  /**
   * value: "revertTracer"
   * @const
   */
  "revertTracer": "revertTracer",
  /**
   * value: "unigramTracer"
   * @const
   */
  "unigramTracer": "unigramTracer",
  /**
   * value: "bigramTracer"
   * @const
   */
  "bigramTracer": "bigramTracer",
  /**
   * value: "trigramTracer"
   * @const
   */
  "trigramTracer": "trigramTracer"
};
var _default = exports["default"] = TracerObject;
