// @klaytn/ethers-ext/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.providers = exports.parseTransaction = exports.KlaytnTxFactory = exports.KlaytnTx = exports.AccountKeyFactory = exports.AccountKey = void 0;
// Pass-through js-ext-core exports
__exportStar(require("@klaytn/js-ext-core/util"), exports);
var js_ext_core_1 = require("@klaytn/js-ext-core");
Object.defineProperty(exports, "AccountKey", { enumerable: true, get: function () { return js_ext_core_1.AccountKey; } });
Object.defineProperty(exports, "AccountKeyFactory", { enumerable: true, get: function () { return js_ext_core_1.AccountKeyFactory; } });
Object.defineProperty(exports, "KlaytnTx", { enumerable: true, get: function () { return js_ext_core_1.KlaytnTx; } });
Object.defineProperty(exports, "KlaytnTxFactory", { enumerable: true, get: function () { return js_ext_core_1.KlaytnTxFactory; } });
Object.defineProperty(exports, "parseTransaction", { enumerable: true, get: function () { return js_ext_core_1.parseTransaction; } });
// ethers-ext classes and functions
__exportStar(require("./accountStore"), exports);
__exportStar(require("./keystore"), exports);
__exportStar(require("./signer"), exports);
// Follow ethers v6 convention like `ethers.JsonRpcProvider`
__exportStar(require("./provider"), exports);
// Follow ethers v5 convention like `ethers.providers.JsonRpcProvider`
const provider_1 = require("./provider");
exports.providers = {
    JsonRpcProvider: provider_1.JsonRpcProvider,
    Web3Provider: provider_1.Web3Provider,
};
